Dockerfile

```diff
@@ -0,0 +1,71 @@
+FROM --platform=linux/amd64 ruby:2.5.0-slim-stretch
+
+ENV NODE_ENV="production" \
+  RAILS_ENV="production" \
+  BUNDLE_DEPLOYMENT="1" \
+  BUNDLE_PATH="/usr/local/bundle" \
+  BUNDLE_WITHOUT="development"
+
+# Install dependencies
+RUN echo "deb http://archive.debian.org/debian stretch main contrib non-free" > /etc/apt/sources.list && \
+  echo "deb http://archive.debian.org/debian-security stretch/updates main contrib non-free" >> /etc/apt/sources.list && \
+  echo 'Acquire::Check-Valid-Until "false";' > /etc/apt/apt.conf.d/99no-check-valid-until && \
+  apt-get update -qq && \
+  apt-get install --no-install-recommends -y \
+  curl build-essential zlib1g-dev git libpq-dev \
+  # Editor
+  vim nano \
+  # For nvm and Node.js
+  libffi-dev python3 python3-pip \
+  # For mimemagic gem
+  shared-mime-info \
+  # Clean up
+  && rm -rf /var/lib/apt/lists/*
+
+# Install Ruby bundler
+RUN bash -c "gem install bundler -v '1.17.3'"
+
+# Install nvm
+ENV NVM_DIR=/usr/local/nvm
+ENV NODE_VERSION=10.24.1
+# Create NVM_DIR directory
+RUN mkdir -p $NVM_DIR
+# Install nvm
+RUN curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
+# Install Node.js and npm
+RUN bash -c "source $NVM_DIR/nvm.sh && nvm install $NODE_VERSION && nvm alias default $NODE_VERSION"
+# Add Node.js and npm to PATH
+ENV PATH="$NVM_DIR/versions/node/v$NODE_VERSION/bin:$PATH"
+# Install Yarn globally using npm
+RUN bash -c "source $NVM_DIR/nvm.sh && npm install -g yarn"
+# Source nvm scripts automatically
+RUN echo 'export NVM_DIR="$NVM_DIR"' >> /etc/bash.bashrc && \
+  echo '[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"' >> /etc/bash.bashrc
+
+# Rails app lives here
+WORKDIR /rails
+
+# Install application gems
+COPY Gemfile Gemfile.lock package.json yarn.lock ./
+RUN bundle install && \
+  rm -rf ~/.bundle/ "${BUNDLE_PATH}"/ruby/*/cache "${BUNDLE_PATH}"/ruby/*/bundler/gems/*/.git && \
+  yarn
+
+# Copy application code
+COPY . .
+
+# Precompiling assets for production
+# TODO: need to change when upgrade Rails.
+RUN SECRET_KEY_BASE=dummy ./bin/rails assets:precompile
```

Comment:
Is `SECRET_KEY_BASE` a placeholder for future rails here? Or is it supposed to be set with a `.env.staging` file - in which case doesn't it get overridden by `dummy` here?

Reply:
SECRET_KEY_BASE is essential in staging and production in Rails, it use to encrypt the cookie/session and password (but not devise).

In the old days, cap deploy is basiclly SSH into the server and run the assets precompile which the .env is there, all good, but actually not really using the it (assets don't need that secret at all).

When we use docker nowadays, we build the image without any secret, but due to the rails stack it still need that to make it run. Newer rails come with a new env for that in [Dockerfile](https://railsdiff.org/7.0.8.7/7.1.5.1#diff-466a28a0e93935ce250159682062e5a94698a3d8):
```
RUN SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile
```
But it is not available in the rails version using in this project.

In my knowledge `RUN SECRET_KEY_BASE=dummy ./bin/rails assets:precompile` inline environment variable only available in that line, not retain in other docker command when build, as well as the container.

There is one SECRET_KEY_BASE in .env in staging and production server. And later if we deploy using kamal, we need to set that in kamal config too. You may find some newer project do not need that env because the value is stored in Rails credential. However this project still using secrets.yml with ENV, thats why.

Reply:
Oh it is me that is the dummy, I got it confused with RAILS_MASTER_KEY. Yep, I agree this is safe.

Reply:
One more thing we need to keep it in mind about assets.

In a perfect setup, the assets should upload to CDN instead of hosting locally.
If we host locally, it should put in a share folder (cap shared/kamal host volume), so that the out-dated assets (from previous releases) still accessable by the end-user.

All assets is build with a checksum in the file name, to ensure the user always loading the latest assets.
However in some scenario, end-user still need to access some old/out-dated assets. For example, if we use an image in the email. The email content point to the old image which may no longer avaiable in the latest release. If we not setup the assets folder properly, that image in the email will be broken.

Good news is, seems all our email in WCMC is simple text, so we don't need to worry about this.